src/middleware/compression.rs

```diff
@@ -0,0 +1,229 @@
+pub struct Compression {
+    default_encoding: Encoding,
+    settings: EncodeSettings,
+}
+
+impl Default for Compression {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Compression {
+    pub fn new() -> Self {
+        Compression {
+            default_encoding: Encoding::Identity,
+            settings: Default::default(),
+        }
+    }
+
+    pub fn with_default(default_encoding: Encoding) -> Self {
+        Compression {
+            default_encoding,
+            settings: Default::default(),
+        }
+    }
+
+    fn parse_encoding(&self, s: &str) -> Result<Encoding, Error> {
+        match s {
+            "gzip" => Ok(Encoding::Gzip),
+            "deflate" => Ok(Encoding::Deflate),
+            "identity" => Ok(Encoding::Identity),
+            "br" => Ok(Encoding::Deflate),
+            "*" => Ok(self.default_encoding),
+            _ => Err(Error::from(StatusCode::UNSUPPORTED_MEDIA_TYPE)),
+        }
+    }
+
+    fn preferred_encoding(&self, headers: &HeaderMap) -> Result<Encoding, Error> {
+        let mut choices: Vec<(Encoding, f32)> = Vec::new();
+        for hval in headers.get_all(ACCEPT_ENCODING).iter() {
+            for cap in RE_ACCEPT_ENCODING.captures_iter(hval.to_str().unwrap()) {
+                match self.parse_encoding(cap.name("a").unwrap().as_str()) {
+                    Ok(encoding) => match cap.name("q") {
+                        Some(qval) => {
+                            let qval = qval.as_str().parse::<f32>().unwrap();
+                            if qval == 0. && encoding == Encoding::Identity {
+                                return Err(Error::from(StatusCode::NOT_ACCEPTABLE));
+                            } else if qval > 0. {
+                                choices.push((encoding, qval))
+                            }
+                        }
+                        None => choices.push((encoding, 1.0)),
+                    },
+                    Err(e) => continue, // ignore encodings we can't perform
+                }
+            }
+        }
+
+        choices.sort_by(|(_, q1), (_, q2)| q2.partial_cmp(&q1).unwrap());
+
+        if !choices.is_empty() {
+            Ok(choices[0].0)
+        } else {
+            Ok(Encoding::Identity)
+        }
+    }
+    /// Consumes the response and returns an encoded version of it.
+    fn encode(&self, mut res: Response, encoding: Encoding) -> Response {
+        res.headers_mut()
+            .append(CONTENT_ENCODING, encoding.header_value());
+        match encoding {
+            Encoding::Gzip => {
+                // FIXME: currently does not encode anything yet
+                let mut body = res.body_mut();
+                let stream = body.map(|b| match b {
+                    Ok(b) => {
+                        let reader = b.into_buf().reader();
+                        let mut buffer: Vec<u8> = Vec::new();
+                        let mut encoder = GzEncoder::new(reader, self.settings.gzip_level);
```

Comment:
`GzEncoder` should be created once outside of closure, and then used

Also use `flate2::write::GzEncoder::new(Vec::new()...`